### PR TITLE
Implement filtering in Log V2 experience

### PR DESCRIPTION
**Pull request by CoderDake**
https://github.com/flutter/devtools/assets/1386322/1248d93d-c040-4a2d-8810-8dc1e75f6757

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 1]
```diff
@@ -5,25 +5,49 @@
...
+final _verboseFlutterFrameworkLogKinds = [
+  FlutterEvent.firstFrame,
+  FlutterEvent.frameworkInitialization,
+  FlutterEvent.frame,
+  FlutterEvent.imageSizesForFrame,
+];
+
+final _verboseFlutterServiceLogKinds = [
+  FlutterEvent.serviceExtensionStateChanged,
+];
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:32]: nit: these can both be Sets

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 2]
```diff
@@ -35,15 +59,17 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
     );
+
+    _retentionLimit = preferences.logging.retentionLimit.value;
+    subscribeToFilterChanges();
   }
 
-  final _logs = ListQueue<LogDataV2>();
-  final _filteredLogs = ListQueue<LogDataV2>();
+  final _logs = ListQueue<_LogEntry>();
+  final _filteredLogs = ListQueue<_FilteredLogEntry>();
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:68]: can you add a comment explaining why we are storing a secondary data set for filtered logs in addition to the `filteredData` provided by the FilterControllerMixin

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 3]
```diff
@@ -78,14 +161,19 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
   set tableWidth(double width) {
     if (width != _tableWidth) {
       _tableWidth = width;
-      cachedHeights.clear();
-      cachedOffets.clear();
+      for (final e in _logs) {
+        e.height = null;
+      }
+      for (final e in _filteredLogs) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:167]: nit: use `log` instead of `e`

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 4]
```diff
@@ -100,23 +188,28 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
   void add(LogDataV2 log) {
-    // TODO(danchevalier): ensure that search and filter lists are updated here.
-
-    _logs.add(log);
-    _filteredLogs.add(log);
-
+    final newEntry = _LogEntry(log);
+    _logs.add(newEntry);
+    getLogHeight(
+      _logs.length - 1,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:194]: no trailing comma when dart code fits on one line

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 5]
```diff
@@ -134,22 +227,77 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
+  double getLogHeight(int index) {
+    final entry = _logs.elementAt(index);
+    final cachedHeight = entry.height;
+    if (cachedHeight != null) return cachedHeight;
+    final height = LoggingTableRow.calculateRowHeight(
+      entry.log,
+      _tableWidth,
+    );
+    entry.height = height;
+    return height;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:239]: nit: this can be condensed:
```
return entry.height ??= LoggingTableRow.calculateRowHeight(
      entry.log,
      _tableWidth,
    );
```

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 6]
```diff
@@ -134,22 +227,77 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
     if (cachedHeight != null) return cachedHeight;
 
-    return cachedHeights[index] = LoggingTableRow.calculateRowHeight(
-      _logs.elementAt(index),
+    final height = LoggingTableRow.calculateRowHeight(
+      filteredLog.logEntry.log,
       _tableWidth,
     );
+    filteredLog.logEntry.height = height;
+    return height;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:253]: same comment here

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 7]
```diff
@@ -134,22 +227,77 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
+class _LogEntry {
+  _LogEntry(this.log);
+  final LogDataV2 log;
+  double? height;
+}
+
+class _FilteredLogEntry {
+  _FilteredLogEntry(this.logEntry);
+  final _LogEntry logEntry;
+  double? offset;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:302]: do we need an entire wrapper class for these? Can't we just keep a map to lookup offsets for each _LogEntry?

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_table_v2.dart [hunk 8]
```diff
@@ -66,16 +58,15 @@ class _LoggingTableV2State extends State<LoggingTableV2> {
           children: [
             Expanded(
               child: DevToolsClearableTextField(
-                labelText: 'Filter',
-                onSubmitted: (value) {},
+                labelText: 'Search',
               ),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_table_v2.dart:62]: Can you use `SearchField` here instead? Fine to do as part of a separate PR, but can you add a TODO?

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 9]
```diff
@@ -463,3 +465,156 @@ extension PatternListExtension on List<Pattern> {
...
+
+  @override
+  void initState() {
+    super.initState();
+    queryTextFieldController = TextEditingController(
+      text: widget.controller.activeFilter.value.queryFilter.query,
+    );
+    useRegExp = widget.controller.useRegExp.value;
+    addAutoDisposeListener(widget.controller.useRegExp, () {
+      useRegExp = widget.controller.useRegExp.value;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:493]: this needs to be wrapped in a set state call for the widget to be rebuilt

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 10]
```diff
@@ -463,3 +465,156 @@ extension PatternListExtension on List<Pattern> {
...
+            );
+          },
+          isFilterActive: false,
+        ),
+      ],
+    );
+  }
+}
+
+class FilterOptionsDialog<T> extends StatefulWidget {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:555]: we already have a class FilterDialog. Why are we creating a new one?

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 11]
```diff
@@ -463,3 +465,156 @@ extension PatternListExtension on List<Pattern> {
...
+  }
+
+  @override
+  void dispose() {
+    queryTextFieldController.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:504]: this seems to be duplicating a bunch of existing code. Can we modify our existing Filter functionality to fit this use case instead of creating two very similar widgets?

File: packages/devtools_app/lib/src/shared/ui/filter.dart (hunk 10 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/shared/ui/filter.dart:555]: Ah thanks for reminding me I need to rename this.
The filtering field used to be inside this dialog, the new implementation offers a field that has a button to open this dialog, which contains the toggle-able filters.

File: packages/devtools_app/lib/src/shared/ui/filter.dart (hunk 11 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/shared/ui/filter.dart:504]: This is possible but could this be left as an improvement to do in the future? In the name of moving fast to get the logging v2 change through I think that touching things outside of the logging_v2 directory gets into much slower and more rigorous territory.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 12]
```diff
@@ -98,25 +195,272 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
+        // Get info about the stackTrace object.
+        if (_isNotNull(stackTrace)) {
+          result += '\n\n${_valueAsString(stackTrace)}';
+        }
+
+        return result;
+      };
+    }
+
+    const severeIssue = 1000;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:424]: Where is this from? Is it arbitrary?

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 13]
```diff
@@ -98,25 +195,272 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
+    final isError = level != null && level >= severeIssue ? true : false;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:425]: `final isError = level != null && level >= severeIssue;` does the same thing.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 14]
```diff
@@ -98,25 +195,272 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
-    _logs.add(log);
-    _filteredLogs.add(log);
+    if (!_filterCallback(newEntry)) {
+      // Only add the log to filtered logs if it matches the filter.
+      return;
+    }
+    _filteredLogs.add(_FilteredLogEntry(newEntry));
 
-    _trimOneOutOfRetentionLog();
+    // TODO(danchevalier): Calculate the new offset here
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:452]: Are you planning on doing this in a follow up or this PR?

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 15]
```diff
@@ -134,13 +478,24 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
     throw Exception('Implement this when needed');
   }
 
+  double getLogHeight(int index) {
+    final entry = _logs.elementAt(index);
+    final cachedHeight = entry.height;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:483]: I don't think these two lines are necessary. If `entry.height != null`, the conditional assignment below doesn't happen and `entry.height` is still returned.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 16]
```diff
@@ -150,6 +505,46 @@ class LoggingTableModel extends ChangeNotifier with DisposerMixin {
...
+        ],
+        ToggleFilter<LogDataV2>(
+          name: 'Hide garbage collection logs',
+          includeCallback: (log) => !log.kind.caseInsensitiveEquals(_gcLogKind),
+          enabledByDefault: true,
+        ),
+      ];
+}
+
+class _LogEntry {
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:540]: Documentation for these classes?

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_table_v2.dart [hunk 17]
```diff
@@ -34,28 +37,17 @@ class LoggingTableV2 extends StatefulWidget {
...
   String lastSearch = '';
   double maxWidth = 0.0;
 
   @override
-  void initState() {
-    super.initState();
-    // On web, disable the browser's context menu since this example uses a custom
-    // Flutter-rendered context menu.
-    if (kIsWeb) {
-      unawaited(BrowserContextMenu.disableContextMenu());
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_table_v2.dart:49]: Is this no longer relevant?

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 18]
```diff
@@ -463,3 +465,167 @@ extension PatternListExtension on List<Pattern> {
...
+  late bool useRegExp;
+
+  @override
+  void initState() {
+    super.initState();
+    queryTextFieldController = TextEditingController(
+      text: widget.controller.activeFilter.value.queryFilter.query,
+    );
+    useRegExp = widget.controller.useRegExp.value;
+    addAutoDisposeListener(widget.controller.useRegExp, () {
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:500]: Is `FilterControllerMixin<T>.useRegExp` a `ValueListenable`? If so, we should just use a `ValueListenableBuilder` in the build method and not register a listener here.

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 19]
```diff
@@ -463,3 +465,167 @@ extension PatternListExtension on List<Pattern> {
...
+    return StateUpdateDialog(
+      title: 'Filter Settings',
+      onApply: _applyFilterChanges,
+      onCancel: _restoreOldValues,
+      onResetDefaults: _resetFilters,
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: [
+          for (final toggleFilter in widget.controller._toggleFilters) ...[
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:605]: I don't think you need the list spread here.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart (hunk 12 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:424]: Yeah, a lot of chunks of the code are copied from the original logging controller

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart (hunk 13 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:425]: Oops I accidentally included these changes here, but they were supposed to be on the next PR. Removing them.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart (hunk 14 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:452]: yessir offsets will only really be useful with search behaviour so I'm going to nail those down in that PR.

File: packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart [hunk 20]
```diff
@@ -537,14 +295,28 @@ class LoggingTableModel extends DisposableController
       ];
 }
 
+/// A class for holding a [LogDataV2] and it's current estimated [height].
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/logging/logging_screen_v2/logging_model.dart:298]: Ubernit: 'its'

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 21]
```diff
@@ -515,29 +508,34 @@ class _StandaloneFilterFieldState<T> extends State<StandaloneFilterField<T>>
...
+                labelText: 'Filter',
+                controller: queryTextFieldController,
+                additionalSuffixActions: [
+                  DevToolsToggleButton(
+                    icon: Codicons.regex,
+                    message: 'Use regular expressions',
+                    outlined: false,
+                    isSelected: useRegExp,
+                    onPressed: () => setState(() {
+                      widget.controller.useRegExp.value = !useRegExp;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:525]: Won't this trigger two rebuilds? One when you update the `ValueListenable` and one when `setState` is called. Maybe the `setState` call can be removed?

**Review by bkonyi — APPROVED** [review]

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 22]
```diff
@@ -130,12 +130,12 @@ mixin FilterControllerMixin<T> on DisposableController
...
 
-/// Dialog to manage filter settings.
+/// Dialog to manage toggleable filter settings.
 ///
 /// This dialog interacts with a [FilterControllerMixin] to manage and preserve
-/// the filter state managed by the dialog.
-class FilterDialog<T> extends StatefulWidget {
-  FilterDialog({
+/// the toggleable filter state managed by the dialog.
+class ToggleFilterDialog<T> extends StatefulWidget {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:137]: We should undo this rename since this isn't only for toggle filters. This dialog also contains a query filter by default.

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 23]
```diff
@@ -552,8 +560,10 @@ class _FilterFieldState<T> extends State<FilterField<T>> with AutoDisposeMixin {
   }
 }
 
-class FilterOptionsDialog<T> extends StatefulWidget {
-  FilterOptionsDialog({
+/// A dialog for editing settings for only the toggleFilters associated with a
+/// [FilterControllerMixin].
+class ToggleFilterOptionsDialog<T> extends StatefulWidget {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:565]: can you add a TODO to try to merge this with FilterDialog since it is duplicating a lot of the same functionality?

File: packages/devtools_app/lib/src/shared/ui/filter.dart (hunk 22 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/shared/ui/filter.dart:137]: This was accidental, It is flipped back now :)

File: packages/devtools_app/lib/src/shared/ui/filter.dart (hunk 23 above)
Comment (CoderDake) [reply packages/devtools_app/lib/src/shared/ui/filter.dart:565]: I honestly didn't notice that, flipping it now :)

File: packages/devtools_app/lib/src/shared/ui/filter.dart [hunk 24]
```diff
@@ -128,10 +130,10 @@ mixin FilterControllerMixin<T> on DisposableController
   }
 }
 
-/// Dialog to manage filter settings.
+/// Dialog to manage toggleable filter settings.
 ///
 /// This dialog interacts with a [FilterControllerMixin] to manage and preserve
-/// the filter state managed by the dialog.
+/// the toggleable filter state managed by the dialog.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/ui/filter.dart:136]: remove docs updates

**Review by kenzieschmoll — APPROVED** [review]
one more comment and lgtm
